modules/packages/MPI.chpl

```diff
@@ -260,6 +265,20 @@ module MPI {
      processes (world size)
      */
   proc initialize() {
+    // If we are running using the uGNI layer, then the following hack
+    // appears to be necessary in order to run MPI, as well as Chapel
+    // See : https://hpcrdm.lbl.gov/pipermail/upc-users/2014-May/002061.html
+    if isUGNI {
+      coforall loc in Locales do on loc {
+          // This must be done on all locales!
+          var pmiGniCookie = C_Env.getenv("PMI_GNI_COOKIE") : string;
+          if !pmiGniCookie.isEmptyString() {
+            C_Env.setenv("PMI_GNI_COOKIE",("%i".format(pmiGniCookie:int+1)).c_str(),1);
+          }
```

Comment:
I think this is okay, but I'm not positive.  I believe that the "proper" way to do this would be to select an otherwise unused instance ID value (the first argument to `GNI_CdmCreate()`) instead. However, for that to work you'd have to have cooperation among all the callers of `GNI_CdmCreate()` and here you're specifically dealing with a situation where the two users of uGNI (MPI and the Chapel comm layer) don't know about each other.  So, ~~lying about~~ adjusting the cookie value may be the only practical alternative.  (Also, at the limit of its scaling the Chapel ugni comm layer will use all values of the the instance ID itself.  So in that situation there would not be any "free" ones for MPI to use even if you could get it to pick a different one than the 0 (zero) it's likely defaulting to.)